### PR TITLE
Avoid dereferencing endpoint twice on the deletion of a service

### LINKED ISSUE #93229: Windows Kubeproxy has Connectivity issues when one of the multiple services referencing a same endpoint is deleted or updated
**What happened**:
Two services referring the same endpoint. One of the services was deleted. This brought down the connectivity of the other service.

**What you expected to happen**:
The connectivity to service should continue to work

**How to reproduce it (as minimally and precisely as possible)**:
1. Create two services referencing same endpoint
2. Delete one of the services
3. The connectivity is broken to the other service

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by sbangari**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
When a service is deleted, the corresponding remote endpoints are dereferenced. But the map holding the list of remote endpoints created for this Service is not updated. This causes the remote endpoints to become stale and they are attempted to be cleaned up a second time later causing a second dereference.

In a scenario where the remote endpoints are shared across two services, the second dereference causes the remote endpoint to be deleted from HNS.

**Which issue(s) this PR fixes**:

Fixes #93229

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Review by elweb9858 — APPROVED** [review]
lgtm. Have you thought about writing a test for this scenario? Maybe creating two services referencing the same remote endpoints, deleting one of the services, and verifying the endpoints still exist.

**Comment by elweb9858** [comment]
also are you aiming to get this bug fix into v1.19? fyi @daschott

File: pkg/proxy/winkernel/proxier.go [hunk 1]
```diff
@@ -279,6 +279,9 @@ func (proxier *Proxier) onServiceMapChange(svcPortName *proxy.ServicePortName) {
 
 		klog.V(3).Infof("Updating existing service port %q at %s:%d/%s", svcPortName, svcInfo.ClusterIP(), svcInfo.Port(), svcInfo.Protocol())
 		svcInfo.cleanupAllPolicies(proxier.endpointsMap[*svcPortName])
+		// Delete the endpoint map entry for this service, since the
+		// endpoints are already cleanedup
+		delete(proxier.endpointsMap, *svcPortName)
```
Comment (kumarvin123) [inline pkg/proxy/winkernel/proxier.go:284]: This is a  problem. The endpoints might be still there and the Svc might have been updated. We clean up all policies assuming they might be stale. This code is unconditionally deleting all endpoints from the map.

Comment (sbangari) [reply pkg/proxy/winkernel/proxier.go:284]: I opened a new PR for this issue. #93368

**Comment by sbangari** [comment]
> lgtm. Have you thought about writing a test for this scenario? Maybe creating two services referencing the same remote endpoints, deleting one of the services, and verifying the endpoints still exist.

@elweb9858   I opened a new PR for this issue and which also has new tests added for the failing scenario. #93368

**Pull request closed without merging**